providers/asrockrack/helpers.go

```diff
@@ -0,0 +1,408 @@
+package asrockrack
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"mime/multipart"
+	"net/http"
+	"net/http/httputil"
+	"os"
+	"path"
+
+	"github.com/bmc-toolbox/bmclib/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+// API session setup response payload
+type loginSession struct {
+	CSRFToken         string `json:"csrftoken,omitempty"`
+	Privilege         int    `json:"privilege,omitempty"`
+	RACSessionID      int    `json:"racsession_id,omitempty"`
+	ExtendedPrivilege int    `json:"extendedpriv,omitempty"`
+}
+
+// Firmware info endpoint response payload
+type firmwareInfo struct {
+	BMCVersion       string `json:"BMC_fw_version"`
+	BIOSVersion      string `json:"BIOS_fw_version"`
+	MEVersion        string `json:"ME_fw_version"`
+	MicrocodeVersion string `json:"Micro_Code_version"`
+	CPLDVersion      string `json:"CPLD_version"`
+	CMVersion        string `json:"CM_version"`
+	BPBVersion       string `json:"BPB_version"`
+	NodeID           string `json:"Node_id"`
+}
+
+// Payload to preseve config when updating the BMC firmware
+type preserveConfig struct {
+	FlashStatus     int `json:"flash_status"` // 1 = full firmware flash, 2 = section based flash, 3 - version compare flash
+	PreserveConfig  int `json:"preserve_config"`
+	PreserveNetwork int `json:"preserve_network"`
+	PreserveUser    int `json:"preserve_user"`
+}
+
+// Firmware flash progress
+//{ "id": 1, "action": "Flashing...", "progress": "12% done         ", "state": 0 }
+//{ "id": 1, "action": "Flashing...", "progress": "100% done", "state": 0 }
+type upgradeProgress struct {
+	ID       int    `json:"id,omitempty"`
+	Action   string `json:"action,omitempty"`
+	Progress string `json:"progress,omitempty"`
+	State    int    `json:"state,omitempty"`
+}
+
+// BIOS upgrade commands
+// 2 == configure
+// 3 == apply upgrade
+type biosUpdateAction struct {
+	Action int `json:"action"`
+}
+
+// 1 Set BMC to flash mode and prepare flash area
+// at this point all logged in sessions are terminated
+// and no logins are permitted
+func (a *ASRockRack) setFlashMode() error {
+
+	endpoint := "api/maintenance/flash"
+
+	_, statusCode, err := a.queryHTTPS(endpoint, "PUT", nil, nil)
+	if err != nil {
+		return err
+	}
+
+	if statusCode != 200 {
+		return fmt.Errorf("Non 200 response: %d", statusCode)
+	}
+
+	a.resetRequired = true
+
+	return nil
+}
+
+// 2 Upload the firmware file
+func (a *ASRockRack) uploadFirmware(endpoint string, filePath string) error {
+
+	// setup a buffer for our multipart form
+	var form bytes.Buffer
+	w := multipart.NewWriter(&form)
+
+	// create form data from update image
+	fwWriter, err := w.CreateFormFile("fwimage", path.Base(filePath))
```

Comment:
looks like the `filename` is [optional](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition#directives).  With that, `uploadFirmware` could take an `io.Reader` or `io.ReadCloser`.

Reply:
cool, fixed in f47ca2daaa37506b9418202a2eb5bbb755fb29f3